crypto/s2n_ecc_evp.c

```diff
@@ -0,0 +1,160 @@
+};
+#endif
+
+/* IANA values can be found here:
+ * https://tools.ietf.org/html/rfc8446#appendix-B.3.1.4 */
+/* Share sizes are described here:
+ * https://tools.ietf.org/html/rfc8446#section-4.2.8.2 and include the extra
+ * "legacy_form" byte */
+const struct s2n_ecc_named_curve *const s2n_ecc_evp_supported_curves[] = {
+    &s2n_ecc_curve_secp256r1,
+    &s2n_ecc_curve_secp384r1,
+#if S2N_IS_X25519_SUPPORTED
+    &s2n_ecc_curve_x25519,
+#endif
+};
+
+#if S2N_IS_X25519_SUPPORTED
+static int s2n_ecc_evp_generate_key_x25519(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey);
+#endif
+
+static int s2n_ecc_evp_generate_key_nist_curves(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey);
+static int s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey);
+static int s2n_ecc_evp_compute_shared_secret(EVP_PKEY *own_key, EVP_PKEY *peer_public, struct s2n_blob *shared_secret);
+
+#if S2N_IS_X25519_SUPPORTED
+static int s2n_ecc_evp_generate_key_x25519(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey) {
+
+    DEFER_CLEANUP(EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(named_curve->libcrypto_nid, NULL),
+                  EVP_PKEY_CTX_free_pointer);
+    S2N_ERROR_IF(pctx == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    GUARD_OSSL(EVP_PKEY_keygen_init(pctx), S2N_ERR_ECDHE_GEN_KEY);
+    GUARD_OSSL(EVP_PKEY_keygen(pctx, evp_pkey), S2N_ERR_ECDHE_GEN_KEY);
+    S2N_ERROR_IF(evp_pkey == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    return 0;
+}
+#endif
+
+static int s2n_ecc_evp_generate_key_nist_curves(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey) {
+
+    DEFER_CLEANUP(EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL), EVP_PKEY_CTX_free_pointer);
+    S2N_ERROR_IF(pctx == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    GUARD_OSSL(EVP_PKEY_paramgen_init(pctx), S2N_ERR_ECDHE_GEN_KEY);
+    GUARD_OSSL(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, named_curve->libcrypto_nid), S2N_ERR_ECDHE_GEN_KEY);
+
+    DEFER_CLEANUP(EVP_PKEY *params = NULL, EVP_PKEY_free_pointer);
+    GUARD_OSSL(EVP_PKEY_paramgen(pctx, &params), S2N_ERR_ECDHE_GEN_KEY);
+    S2N_ERROR_IF(params == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    DEFER_CLEANUP(EVP_PKEY_CTX *kctx = EVP_PKEY_CTX_new(params, NULL), EVP_PKEY_CTX_free_pointer);
+    S2N_ERROR_IF(kctx == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    GUARD_OSSL(EVP_PKEY_keygen_init(kctx), S2N_ERR_ECDHE_GEN_KEY);
+    GUARD_OSSL(EVP_PKEY_keygen(kctx, evp_pkey), S2N_ERR_ECDHE_GEN_KEY);
+    S2N_ERROR_IF(evp_pkey == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    return 0;
+}
+
+static int s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey) {
+#if S2N_IS_X25519_SUPPORTED
+    if (named_curve->libcrypto_nid == NID_X25519) {
+        return s2n_ecc_evp_generate_key_x25519(named_curve, evp_pkey);
+    }
+#endif
+    if (named_curve->libcrypto_nid == NID_X9_62_prime256v1 || named_curve->libcrypto_nid == NID_secp384r1) {
```

Comment:
NitPick: It might be worth capturing if a curve is "nist curve" (and can be generated with `s2n_ecc_evp_generate_key_nist_curves`) as a boolean in the `s2n_ecc_named_curve` structure. I don't like haven't an explicit list of curves in the method since it doesn't really care what the curve is, just what method should be used.

For that matter, maybe the generating method should be an element in the structure proper? At that point this method becomes something like:
```
  return named_curve->generate_own_key(named_curve, evp_pkey);
```

Reply:
Agreed on not explicitly listing the curves in the method, I like your idea of adding generate_own_key as a part of the structure, it makes a clearer design.